File: src/portfolio_manager/repositories/holding_repository.py
"""Holding repository for database operations."""

from collections import defaultdict
from decimal import Decimal
from uuid import UUID, uuid4

from portfolio_manager.core.time import now_kst

from portfolio_manager.models import Holding
from portfolio_manager.services.database import HoldingModel, db


class HoldingRepository:
    """Repository for Holding database operations."""

    def create(self, account_id: UUID, stock_id: UUID, quantity: Decimal) -> Holding:
        """Create a new holding."""
        now = now_kst()
        row = HoldingModel.create(
            id=uuid4(),
            account=account_id,
            stock=stock_id,
            quantity=quantity,
            created_at=now,
            updated_at=now,
        )
        return self._to_domain(row)

    def list_by_account(self, account_id: UUID) -> list[Holding]:
        """List holdings for a specific account."""
        return [
            self._to_domain(row)
            for row in HoldingModel.select().where(HoldingModel.account == account_id)
        ]

    def delete_by_account(self, account_id: UUID) -> None:
        """Delete holdings for a specific account."""
        HoldingModel.delete().where(HoldingModel.account == account_id).execute()

    def delete(self, holding_id: UUID) -> None:
        """Delete a holding by ID."""
        HoldingModel.delete().where(HoldingModel.id == holding_id).execute()

    def update(self, holding_id: UUID, quantity: Decimal) -> Holding:
        """Update a holding quantity by ID."""
        now = now_kst()
        HoldingModel.update(quantity=quantity, updated_at=now).where(
            HoldingModel.id == holding_id
        ).execute()
        row = HoldingModel.get_by_id(holding_id)
        return self._to_domain(row)

    def bulk_update_by_account(
        self, account_id: UUID, updates: list[tuple[UUID, Decimal]]
    ) -> list[Holding]:
        """Update multiple holdings for one account atomically."""
        if not updates:
            return []

        holding_ids = [hid for hid, _ in updates]
        if len(set(holding_ids)) != len(holding_ids):
            raise ValueError("duplicate holding_ids are not allowed")

        for _, qty in updates:
            if qty <= 0:
                raise ValueError("quantity must be greater than zero")

        # Verify all holdings belong to the account
        existing = {
            row.id: row
            for row in HoldingModel.select().where(
                (HoldingModel.account == account_id)
                & (HoldingModel.id.in_(holding_ids))
            )
        }
        if len(existing) != len(holding_ids):
            raise ValueError("all holdings must belong to account")

        now = now_kst()
        results = []
        with db.atomic():
            for holding_id, quantity in updates:
                HoldingModel.update(quantity=quantity, updated_at=now).where(
                    HoldingModel.id == holding_id
                ).execute()

            for holding_id, _ in updates:
                row = HoldingModel.get_by_id(holding_id)
                results.append(self._to_domain(row))

        return results

    def get_aggregated_holdings_by_stock(self) -> dict[UUID, Decimal]:
        """Get aggregated holdings by stock across all accounts."""
        rows = HoldingModel.select(HoldingModel.stock, HoldingModel.quantity)

        aggregated: dict[UUID, Decimal] = defaultdict(Decimal)
        for row in rows:
            aggregated[row.stock_id] += Decimal(str(row.quantity))

        return dict(aggregated)

    @staticmethod
    def _to_domain(row: HoldingModel) -> Holding:
        return Holding(
            id=row.id,
            account_id=row.account_id,
            stock_id=row.stock_id,
            quantity=Decimal(str(row.quantity)),
            created_at=row.created_at,
            updated_at=row.updated_at,
        )

File: internal/repositories/holding_repository.go
package repositories

import (
	"context"
	"database/sql"
	"errors"
	"fmt"

	"github.com/kadragon/portfolio-manager/internal/db/sqlc"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/numeric"
	"github.com/kadragon/portfolio-manager/internal/uuidx"
)

// HoldingUpdate pairs a holding ID with the new quantity for bulk updates.
type HoldingUpdate struct {
	ID       uuidx.UUID
	Quantity numeric.Decimal
}

// HoldingRepository provides holding CRUD operations.
type HoldingRepository struct {
	q *sqlc.Queries
}

// NewHoldingRepository builds a HoldingRepository over the given queries handle.
func NewHoldingRepository(q *sqlc.Queries) *HoldingRepository {
	return &HoldingRepository{q: q}
}

// Create inserts a new holding and returns it.
func (r *HoldingRepository) Create(ctx context.Context, accountID, stockID uuidx.UUID, quantity numeric.Decimal) (models.Holding, error) {
	now := ktime.Now()
	row, err := r.q.CreateHolding(ctx, sqlc.CreateHoldingParams{
		ID:        uuidx.New(),
		AccountID: accountID,
		StockID:   stockID,
		Quantity:  quantity,
		CreatedAt: now,
		UpdatedAt: now,
	})
	if err != nil {
		return models.Holding{}, err
	}
	return toHolding(row), nil
}

// ListByAccount returns all holdings for the given account in insertion order.
func (r *HoldingRepository) ListByAccount(ctx context.Context, accountID uuidx.UUID) ([]models.Holding, error) {
	rows, err := r.q.ListHoldingsByAccount(ctx, accountID)
	if err != nil {
		return nil, err
	}
	out := make([]models.Holding, 0, len(rows))
	for _, row := range rows {
		out = append(out, toHolding(row))
	}
	return out, nil
}

// GetByID returns the holding with the given id, or nil if absent.
func (r *HoldingRepository) GetByID(ctx context.Context, id uuidx.UUID) (*models.Holding, error) {
	row, err := r.q.GetHoldingByID(ctx, id)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	h := toHolding(row)
	return &h, nil
}

// Update changes the quantity of a holding.
func (r *HoldingRepository) Update(ctx context.Context, id uuidx.UUID, quantity numeric.Decimal) (models.Holding, error) {
	row, err := r.q.UpdateHolding(ctx, sqlc.UpdateHoldingParams{
		Quantity:  quantity,
		UpdatedAt: ktime.Now(),
		ID:        id,
	})
	if err != nil {
		return models.Holding{}, err
	}
	return toHolding(row), nil
}

// Delete removes the holding with the given id.
func (r *HoldingRepository) Delete(ctx context.Context, id uuidx.UUID) error {
	return r.q.DeleteHolding(ctx, id)
}

// BulkUpdateByAccount validates that all updates belong to the account, then
// applies each update sequentially. Parity with Python: validation completes
// before any write; no transaction needed at this scale.
func (r *HoldingRepository) BulkUpdateByAccount(ctx context.Context, accountID uuidx.UUID, updates []HoldingUpdate) error {
	if len(updates) == 0 {
		return nil
	}

	existing, err := r.q.ListHoldingsByAccount(ctx, accountID)
	if err != nil {
		return err
	}
	existingIDs := make(map[uuidx.UUID]struct{}, len(existing))
	for _, h := range existing {
		existingIDs[h.ID] = struct{}{}
	}
	for _, u := range updates {
		if _, ok := existingIDs[u.ID]; !ok {
			return fmt.Errorf("all holdings must belong to account")
		}
	}

	now := ktime.Now()
	for _, u := range updates {
		if _, err := r.q.UpdateHolding(ctx, sqlc.UpdateHoldingParams{
			Quantity:  u.Quantity,
			UpdatedAt: now,
			ID:        u.ID,
		}); err != nil {
			return err
		}
	}
	return nil
}

func toHolding(row sqlc.Holding) models.Holding {
	return models.Holding{
		ID:        row.ID,
		AccountID: row.AccountID,
		StockID:   row.StockID,
		Quantity:  row.Quantity,
		CreatedAt: row.CreatedAt.Time,
		UpdatedAt: row.UpdatedAt.Time,
	}
}
